lmdeploy/pytorch/models/phi3.py

```diff
@@ -0,0 +1,255 @@
+        ]
+        colwise_split_parallelize_linear_fn(mod, sections, device_mesh)
+
+    def _distribute_partition_fn(self, mod_name: str, mod: nn.Module,
+                                 device_mesh: DeviceMesh):
+        """Distribution partition callback."""
+        if mod_name in ['qkv_proj']:
+            self._distribute_qkv_linear(mod, device_mesh)
+        elif mod_name in ['o_proj']:
+            rowwise_parallelize_linear_fn(mod,
+                                          device_mesh=device_mesh,
+                                          to_local=True)
+
+    @classmethod
+    def _distribute_output_fn(cls, outputs, device_mesh: DeviceMesh):
+        """Distribution output hook."""
+        dist.all_reduce(outputs[0])
+        return outputs
+
+    def _contiguous_batching_forward_impl(
+        self,
+        hidden_states: torch.Tensor,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Tuple[torch.Tensor]] = None,
+        world_size: int = 1,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor],
+               Optional[Tuple[torch.Tensor]]]:
+        """Rewrite implementation of LlamaAttention.forward.
+
+        Add continuous batching support. Add paged attention support. TP
+        support.
+        """
+        context = self.context.context
+        q_start_loc = context.q_start_loc
+        kv_seq_length = context.kv_seq_length
+        q_seq_length = context.q_seq_length
+        block_offsets = context.block_offsets
+        max_q_seq_length = context.max_q_seq_length
+        max_kv_seq_length = context.max_kv_seq_length
+        position_ids_1d = context.position_ids_1d
+
+        num_heads = self.num_heads // world_size
+        num_kv_heads = self.num_key_value_heads // world_size
+        head_dim = self.head_dim
+        hidden_size = num_heads * head_dim
+
+        def __qkv_proj(hidden_states):
+            """qkv proj."""
+            qkv_states = self.qkv_proj(hidden_states)
+            query_states, key_states, value_states = qkv_states.split(
+                (num_heads * head_dim, num_kv_heads * head_dim,
+                 num_kv_heads * head_dim),
+                dim=-1,
+            )
+
+            return query_states, key_states, value_states
+
+        def __rotary_emb_fn(query_states, key_states, value_states):
+            """rotary embedding func."""
+            if not hasattr(context, '_cos'):
+                cos, sin = self.rotary_emb(
+                    value_states,
+                    position_ids=position_ids_1d[None, :],
+                    seq_len=max_kv_seq_length)
+                context._cos = cos
+                context._sin = sin
+            cos = context._cos
+            sin = context._sin
+            query_states, key_states = apply_rotary_pos_emb(
+                query_states,
+                key_states,
+                cos[0],
+                sin[0],
+                position_ids,
+                torch.arange(0,
```

Comment:
the `position_ids_1d` seems different. why not use `position_ids_1d` directly?

Reply:
https://github.com/InternLM/lmdeploy/blob/5852aaa0617aefb72099f3b1405daa9ff7035382/lmdeploy/pytorch/models/phi3.py#L86